### ISSUE TITLE
Auto-unpacking works only if both scale_factor and add_offset are defined but not if only one is given

### ISSUE THREAD

**Issue by GoogleCodeExporter** [sjaaky74/netcdf4-python#86]
```
What steps will reproduce the problem?
1.  I am working with data from the Jason altimeter --- for example 
ftp://anonymous@podaac.jpl.nasa.gov/pub/sea_surface_height/jason/gdr_netcdf_c/da
ta/c100/JA1_GPN_2PcP100_254_20041002_174721_20041002_184332.nc 
2. I do the following to retrieve the latitude coordinates of the measurements 
contained in the record:

In [26]: dataset = 
Dataset('JA2_GPN_2PTP055_067_20100101_015040_20100101_024653', 'r')

In [27]: dataset.variables['lat'].maskandscale
Out[27]: True

In [28]: dataset.variables['lat'][0]
Out[28]: -66146438

In [29]: dataset.variables['lat'].scale_factor
Out[29]: 9.9999999999999995e-07


What is the expected output? What do you see instead?

The correct latitude is:

In [23]: dataset.variables['lat'][0] * dataset.variables['lat'].scale_factor
Out[23]: -66.146438000000003

What version of the product are you using? On what operating system?

I am using netcdf4-python 0.9.4 on Mac OSX 10.6.7.  I am using python 2.7 
installed via homebrew http://mxcl.github.com/homebrew/.

Please provide any additional information below.

The documentation for set_auto_maskandscale(self, maskandscale) available at 
http://netcdf4-python.googlecode.com/svn/trunk/docs/netCDF4.Variable-class.html 
states the following:

If maskandscale is set to True, and the variable has a scale_factor and an 
add_offset attribute, then data read from that variable is unpacked using:

   data = self.scale_factor*data + self.add_offset

So, the observed behavior matches the documentation.  However, I think that the 
unpacking procedure should always be applied with defaults scale_factor = 1 and 
add_offset = 0.
```

Original issue reported on code.google.com by `m.d.but...@gmail.com` on 16 Apr 2011 at 6:56

**Comment by GoogleCodeExporter** [comment]
```
I don't like the idea of making assumptions about scale_factor and add_offset, 
if one or the other is missing.  In that case, you will have to unpack the data 
yourself.
```

Original comment by `whitaker.jeffrey@gmail.com` on 16 Apr 2011 at 7:00
- Changed state: **WontFix**

**Comment by GoogleCodeExporter** [comment]
```
Thank you for the reply.  I agree --- assumed behavior is not a good idea.

```

Original comment by `m.d.but...@gmail.com` on 16 Apr 2011 at 8:53

**Comment by GoogleCodeExporter** [comment]
```
I disagree. The Netcdf User guide (see 
http://www.unidata.ucar.edu/software/netcdf/docs/netcdf.html#Attribute-Conventio
ns) explicitly states that if one of scale_factor or add_offset is present, it 
should be applied. It gives additional instruction (in particular which is to 
be applied first) in case both are present.

Thus, I believe that the present implementation which requires both attributes 
to present before they are applied is violating the rather precise 
recommendation in the netcdf documentation.

Attached is a patch which fixes this (unfortunately for 0.9.3 only - I'll redo 
it for 0.9.4 as soon as possible).
```

Original comment by `Marquard...@gmail.com` on 11 May 2011 at 11:00

Attachments:
- [netCDF4-0.9.3.scale_offset.patch](https://storage.googleapis.com/google-code-attachments/netcdf4-python/issue-86/comment-3/netCDF4-0.9.3.scale_offset.patch)

**Comment by GoogleCodeExporter** [comment]
```
Here's the patch for 0.9.4) where it of course only netCDF4.pyx is affected. 
The C-version of the file must be re-generated with cython, of course.

One more comment: I would agree that users shall do their own unpacking if 
needed, but unfortunately the default setting (at least in 0.9.3) was to enable 
unpacking for all variables. One therefore has to disable the internal 
unpacking before being able to implement a 'correct' unpacking. It would be 
useful to be able to enable or disable the internal unpacking globally rather 
than on a variable-by-variable basis. IMHO it shouldn't be the default behavior 
anyway as long as the implementation doesn't follow the stated netCDF 
conventions (see the link in my previous comment).
```

Original comment by `Marquard...@gmail.com` on 11 May 2011 at 11:33

Attachments:
- [netCDF4-0.9.4.scale_offset.patch](https://storage.googleapis.com/google-code-attachments/netcdf4-python/issue-86/comment-4/netCDF4-0.9.4.scale_offset.patch)
